cardano-crypto-class/src/Cardano/Crypto/Seed.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+
+-- | Seeds for key generation.
+--
+module Cardano.Crypto.Seed
+  ( Seed
+    -- * Constructing seeds
+  , mkSeedFromBytes
+  , readSeedFromSystemEntropy
+    -- * Using seeds
+  , getBytesFromSeed
+  , runMonadRandomWithSeed
+  , SeedBytesExhausted(..)
+  ) where
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import           Data.ByteArray as BA (convert)
+import           Numeric.Natural (Natural)
+
+import           Control.Exception (Exception(..), throw)
+
+import           Data.Functor.Identity
+import           Control.Monad.Trans.Maybe
+import           Control.Monad.Trans.State
+
+import           Crypto.Random (MonadRandom(..))
+import           Crypto.Random.Entropy (getEntropy)
+
+import           Cardano.Prelude (NoUnexpectedThunks)
+
+
+-- | A seed contains a finite number of bytes, and is used for seeding
+-- cryptographic algorithms including key generation.
+--
+-- This is not itself a PRNG, but can be used to seed a PRNG.
+--
+newtype Seed = Seed ByteString
+  deriving (Show, Eq, Semigroup, Monoid, NoUnexpectedThunks)
+
+
+-- | Construct a 'Seed' deterministically from a number of bytes.
+--
+mkSeedFromBytes :: ByteString -> Seed
+mkSeedFromBytes = Seed
```

Comment:
Should we insist on a minimum number of bytes here? `mkSeedFromBytes ""` is valid but would not be a good seed.